develop/blockchains/custom-blockchains/benchmarking.md

```diff
@@ -1,4 +1,183 @@
+
+- Computational complexity
+- Storage complexity (proof size)
+- Database reads and writes 
+- Hardware specifications
+
+Benchmark testing uses real-world testing to simulate worst-case scenarios for extrinsics. The framework generates a linear model for weight calculation by running multiple iterations with varied parameters. These worst-case weights ensure blocks remain within execution limits, enabling the runtime to maintain throughput under varying loads. Excess fees can be refunded if a call uses fewer resources than expected, offering users a fair cost model.
+  
+Because weight is a generic unit of measurement based on computation time for a specific physical machine, the weight of any function can change based on the specifications of hardware used for benchmark testing. By modeling the expected weight of each runtime function, the blockchain can calculate the number of transactions or system-level calls it can execute within a certain period of time.
+
+Within FRAME, each function call that is dispatched must have a `#[pallet::weight]` annotation that can return the expected weight for the worst-case scenario execution of that function given its inputs:
+
+```rust hl_lines="2"
+#[pallet::call_index(0)]
+#[pallet::weight(T::WeightInfo::do_something())]
+pub fn do_something(origin: OriginFor<T>, bn: u32) -> DispatchResultWithPostInfo { ... }
+```
+
+The `WeightInfo` file is automatically generated during benchmark testing. Based on these tests, this file provides accurate weights for each extrinsic.
+
+## Benchmark Process
+
+Benchmark testing a pallet involves the following steps: 
+
+1. Creating a `benchmarking.rs` file within your pallet's structure
+2. Writing a benchmarking test for each extrinsic
+3. Executing the benchmarking tool to calculate weights based on performance metrics
+
+The benchmarking tool runs multiple iterations to model worst-case execution times and determine the appropriate weight. By default, the benchmark testing pipeline is deactivated. To activate it, compile your runtime with the `runtime-benchmarks` feature flag.
+
+### Prepare Your Environment
+
+Before writing benchmark tests, you need to ensure the `frame-benchmarking` crate is included in your pallet's `Cargo.toml` similar to the following:
+
+```toml title="Cargo.toml"
+frame-benchmarking = { version = "37.0.0", default-features = false }
+```
+
+You must also ensure that you add the `runtime-benchmarks` feature flag as follows under the `[features]` section of your pallet's `Cargo.toml`:
+
+```toml title="Cargo.toml"
+runtime-benchmarks = [
+    "frame-benchmarking/runtime-benchmarks",
+    "frame-support/runtime-benchmarks",
+    "frame-system/runtime-benchmarks",
+    "sp-runtime/runtime-benchmarks",
+]
+```
+
+Lastly, ensure that `frame-benchmarking` is included in `std = []`: 
+
+```toml title="Cargo.toml"
+std = [
+    # ...
+    "frame-benchmarking?/std",
+    # ...
+]
+```
```

Comment:
These snippets can be moved to a single snippet in the snippets directory. That way, we can ensure that the formatter is run on them